tests/end_to_end_tests/verify.py

```diff
@@ -0,0 +1,309 @@
+# Copyright 2014-2022 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import platform
+import re
+import time
+from urllib.parse import quote_plus, urlencode
+from typing import Callable, List, Any
+
+import requests
+
+from tests.end_to_end_tests.tools import AgentCommander
+
+
+log = logging.getLogger(__name__)
+
+
+AGENT_LOG_LINE_TIMESTAMP = r"\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d+Z"
+
+
+def preprocess_agent_log_messages(content: str):
+
+    lines = content.splitlines(keepends=True)
+
+    if not lines:
+        return []
+
+    # Remove last line if it's incomplete.
+    if not lines[-1].endswith(os.linesep):
+        lines.pop(-1)
+
+    lines = [line.strip() for line in lines]
+
+    messages = []
+    for line in lines:
+        line = line.strip()
+        if re.match(rf"{AGENT_LOG_LINE_TIMESTAMP} .*", line):
+            messages.append((line, []))
+        else:
+            # If line does not start with agent log preamble, then it has to be a multiline message
+            # or error traceback, so we all those additional lines to previous message in additional list.
+            messages[-1][1].append(line)
+
+    return messages
+
+
+def check_agent_log_for_errors(content: str):
+
+    messages = preprocess_agent_log_messages(content=content)
+    error_line_pattern = re.compile(rf"{AGENT_LOG_LINE_TIMESTAMP} (ERROR|CRITICAL) .*")
+    for message, additional_lines in messages:
+        # error is detected, normally fail the test, but also need to check for some particular error
+        # that we may want pass.
+
+        whole_error = message + "\n" + "\n".join(additional_lines)
+        if error_line_pattern.match(message):
+
+            to_fail = True
+
+            # There is an issue with dns resolution on GitHub actions side, so we skip some of the error messages.
+            connection_error_mgs = '[error="client/connectionFailed"] Failed to connect to "https://agent.scalyr.com" due to errno=-3.'
+
+            if connection_error_mgs in messages:
+                # If the traceback that follows after error message contains particular error message,
+                # then we are ok with that.
+                errors_to_ignore = [
+                    "socket.gaierror: [Errno -3] Try again",
```

Comment:
I believe this code is duplicated in another file, right?

If so, it would be good if we could refactor some of it into some common module / similar.

Reply:
Yes, I just got it from the tests/package_tests, but I plan to remove and replace them with these tests. I don't want to get any dependencies from here, since it has many OOP abstractions that only add extra complexity and I really not happy about them now...

Reply:
So to clarify, this will be refactored in a future PR?